server/src/main/java/org/elasticsearch/search/query/SingleThreadCollectorManager.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/** {@link CollectorManager} implementation that assumes it is running non-concurrently. Therefore
+ * {@link #newCollector()} will be called only once. */
+public abstract class SingleThreadCollectorManager implements CollectorManager<Collector, Void> {
+
+    private Collector collector;
+
+    @Override
+    public final Collector newCollector() throws IOException {
+        assert collector == null;
+        collector = getNewCollector();
+        return collector;
+    }
+
+    /** Return a new Collector. This method will be called just once */
+    protected abstract Collector getNewCollector() throws IOException;
+
+    @Override
+    public final Void reduce(Collection<Collector> collectors) throws IOException {
+        assert collectors.size() == 1;
+        Collector collector = collectors.iterator().next();
+        assert this.collector == collector;
+        reduce(collector);
```

Comment:
When I look into the code I don't see any place where it will be useful to return a value. The current search works in a way that we register our collectors / collector managers in the `SearchContext` and we access the results from there. Therefore I don't see the usefulness to return any value and I think it will add so much complexity.  

Reduce will compute the final result and register in the collector manager. We will access those results directly from the collector manager.

Reply:
I understand the complexity of returning a value. My point is not that we should return a value, it's that the reduce method in the single threaded collector manager does nothing and we could remove it?

Reply:
The reduce method is called in lucene even for single threaded execution so I think we should call the reduce method on all our collector managers on the tree. This will allow us to include at any point managers that do stuff (e.g release resources on reduce?)

Reply:
Lucene will always call the reduce method, that's correct. My point is that the single threaded collector manager has nothing to reduce and its reduce method should be a no-op, otherwise it is confusing. The only implementation I've seen in your PR is the aggregation collector manager that does not even read the collector argument. What I was suggesting is to make reduce(List<Collector>) an empty final method, and not require that reduce(Collector) method which looks bogus to me, unless I am missing what that is useful for.